File: backend/app/routers/leaderboard.py
"""Leaderboard — build order B.6. Read-only; there is no client write endpoint (§21)."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Literal

from fastapi import APIRouter, Query
from sqlalchemy import func, select

from app.core.deps import ConsentedUser, Context, DbSession
from app.models.enums import Platform
from app.models.gameplay import LeaderboardEntry
from app.schemas.common import ApiModel
from app.services import analytics

router = APIRouter(prefix="/v1/leaderboard", tags=["leaderboard"])

Window = Literal["all_time", "weekly", "daily"]


class LeaderboardRow(ApiModel):
    rank: int
    display_name: str
    score: int
    level_id: int
    recorded_at: datetime
    is_current_user: bool


class YourRank(ApiModel):
    rank: int | None
    score: int | None
    display_name: str
    level_id: int | None


class LeaderboardResponse(ApiModel):
    window: Window
    rows: list[LeaderboardRow]
    your_rank: YourRank
    total_entries: int
    generated_at: datetime


def _window_start(window: Window) -> datetime | None:
    now = datetime.now(timezone.utc)
    if window == "daily":
        return now.replace(hour=0, minute=0, second=0, microsecond=0)
    if window == "weekly":
        start_of_day = now.replace(hour=0, minute=0, second=0, microsecond=0)
        return start_of_day - timedelta(days=start_of_day.weekday())
    return None


@router.get("", response_model=LeaderboardResponse)
def read_leaderboard(
    user: ConsentedUser,
    db: DbSession,
    context: Context,
    window: Window = Query(default="all_time"),
    limit: int = Query(default=50, ge=1, le=200),
    entry_source: str = Query(default="nav_tab"),
) -> LeaderboardResponse:
    """M-17. Each user's single best entry is ranked, not every attempt.

    Otherwise one player replaying a level would fill the whole board.
    UTC boundaries per backend.timezoneHandling; the client renders local time.
    """
    start = _window_start(window)

    best_per_user = (
        select(
            LeaderboardEntry.user_id.label("user_id"),
            func.max(LeaderboardEntry.score).label("best_score"),
        )
        .where(LeaderboardEntry.revoked.is_(False))
        .group_by(LeaderboardEntry.user_id)
    )
    if start is not None:
        best_per_user = best_per_user.where(LeaderboardEntry.recorded_at >= start)
    best_per_user = best_per_user.subquery()

    ranked = (
        select(LeaderboardEntry)
        .join(
            best_per_user,
            (LeaderboardEntry.user_id == best_per_user.c.user_id)
            & (LeaderboardEntry.score == best_per_user.c.best_score),
        )
        .where(LeaderboardEntry.revoked.is_(False))
        .order_by(LeaderboardEntry.score.desc(), LeaderboardEntry.recorded_at.asc())
        .limit(limit)
    )
    if start is not None:
        ranked = ranked.where(LeaderboardEntry.recorded_at >= start)

    entries = list(db.execute(ranked).scalars().all())

    # Deduplicate: a user can tie their own best score on two attempts.
    seen: set = set()
    rows: list[LeaderboardRow] = []
    for entry in entries:
        if entry.user_id in seen:
            continue
        seen.add(entry.user_id)
        rows.append(
            LeaderboardRow(
                rank=len(rows) + 1,
                display_name=entry.display_name,
                score=entry.score,
                level_id=entry.level_id,
                recorded_at=entry.recorded_at,
                is_current_user=entry.user_id == user.id,
            )
        )

    your_best_query = select(LeaderboardEntry).where(
        LeaderboardEntry.user_id == user.id, LeaderboardEntry.revoked.is_(False)
    )
    if start is not None:
        your_best_query = your_best_query.where(LeaderboardEntry.recorded_at >= start)
    your_best = db.execute(
        your_best_query.order_by(LeaderboardEntry.score.desc())
    ).scalars().first()

    your_rank_value: int | None = None
    if your_best is not None:
        better = select(func.count(func.distinct(LeaderboardEntry.user_id))).where(
            LeaderboardEntry.revoked.is_(False), LeaderboardEntry.score > your_best.score
        )
        if start is not None:
            better = better.where(LeaderboardEntry.recorded_at >= start)
        your_rank_value = int(db.execute(better).scalar_one()) + 1

    total_query = select(func.count(func.distinct(LeaderboardEntry.user_id))).where(
        LeaderboardEntry.revoked.is_(False)
    )
    if start is not None:
        total_query = total_query.where(LeaderboardEntry.recorded_at >= start)
    total = int(db.execute(total_query).scalar_one())

    valid_platforms = {item.value for item in Platform}
    analytics.track(
        db,
        "leaderboard_viewed",
        user=user,
        session_id=context.session_id,
        platform=Platform(context.platform) if context.platform in valid_platforms else None,
        properties={
            "connectivity": "online",
            "entry_source": entry_source,
            "user_rank": your_rank_value,
        },
    )

    return LeaderboardResponse(
        window=window,
        rows=rows,
        your_rank=YourRank(
            rank=your_rank_value,
            score=your_best.score if your_best else None,
            display_name=user.leaderboard_display_name,
            level_id=your_best.level_id if your_best else None,
        ),
        total_entries=total,
        generated_at=datetime.now(timezone.utc),
    )

File: mobile/lib/models/leaderboard.dart
/// design-spec §21: read-only board. Entries are written server-side from
/// validated attempts; there is no client submit endpoint.
enum LeaderboardWindow {
  allTime('all_time'),
  weekly('weekly'),
  daily('daily');

  const LeaderboardWindow(this.wireValue);

  final String wireValue;
}

class LeaderboardRow {
  const LeaderboardRow({
    required this.rank,
    required this.displayName,
    required this.score,
    required this.levelId,
    required this.recordedAt,
    required this.isCurrentUser,
  });

  factory LeaderboardRow.fromJson(Map<String, dynamic> json) => LeaderboardRow(
        rank: (json['rank'] as num).toInt(),
        // Already masked to first name + last initial for under-13 players
        // server-side (design-spec §21 COPPA-06).
        displayName: '${json['display_name'] ?? ''}',
        score: (json['score'] as num?)?.toInt() ?? 0,
        levelId: (json['level_id'] as num?)?.toInt() ?? 0,
        recordedAt: DateTime.tryParse('${json['recorded_at']}')?.toUtc(),
        isCurrentUser: json['is_current_user'] == true,
      );

  final int rank;
  final String displayName;
  final int score;
  final int levelId;
  final DateTime? recordedAt;
  final bool isCurrentUser;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'rank': rank,
        'display_name': displayName,
        'score': score,
        'level_id': levelId,
        'recorded_at': recordedAt?.toIso8601String(),
        'is_current_user': isCurrentUser,
      };
}

class YourRank {
  const YourRank({
    required this.rank,
    required this.score,
    required this.displayName,
    required this.levelId,
  });

  factory YourRank.fromJson(Map<String, dynamic> json) => YourRank(
        rank: (json['rank'] as num?)?.toInt(),
        score: (json['score'] as num?)?.toInt(),
        displayName: '${json['display_name'] ?? ''}',
        levelId: (json['level_id'] as num?)?.toInt(),
      );

  final int? rank;
  final int? score;
  final String displayName;
  final int? levelId;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'rank': rank,
        'score': score,
        'display_name': displayName,
        'level_id': levelId,
      };
}

class LeaderboardPage {
  const LeaderboardPage({
    required this.window,
    required this.rows,
    required this.yourRank,
    required this.totalEntries,
    required this.generatedAt,
  });

  factory LeaderboardPage.fromJson(Map<String, dynamic> json) => LeaderboardPage(
        window: '${json['window'] ?? 'all_time'}',
        rows: ((json['rows'] as List<dynamic>? ?? const <dynamic>[]))
            .map((Object? item) => LeaderboardRow.fromJson(item as Map<String, dynamic>))
            .toList(),
        yourRank: YourRank.fromJson(
          (json['your_rank'] as Map<String, dynamic>?) ?? <String, dynamic>{},
        ),
        totalEntries: (json['total_entries'] as num?)?.toInt() ?? 0,
        generatedAt: DateTime.tryParse('${json['generated_at']}')?.toUtc(),
      );

  final String window;
  final List<LeaderboardRow> rows;
  final YourRank yourRank;
  final int totalEntries;
  final DateTime? generatedAt;

  bool get isEmpty => rows.isEmpty;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'window': window,
        'rows': rows.map((LeaderboardRow row) => row.toJson()).toList(),
        'your_rank': yourRank.toJson(),
        'total_entries': totalEntries,
        'generated_at': generatedAt?.toIso8601String(),
      };
}
